### PR TITLE
error: splitted error_message document, added error message format

### PR DESCRIPTION
**Pull request by Piankero**
## Basics

Check relevant points but **please do not remove entries**.
Do not describe the purpose of this PR in the PR description but:

- [ ] Short descriptions should be in the release notes (added as entry in
      `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add something to the the release notes.**
- [x] Longer descriptions should be in documentation or in design decisions.
- [ ] Describe details of how you changed the code in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, should be in the commit messages.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests
- [ ] I ran all tests locally and everything went fine
- [x] affected documentation is fixed
- [ ] I added code comments, logging, and assertions (see doc/CODING.md)
- [ ] meta data is updated (e.g. README.md of plugins and doc/METADATA.md)

## Review

Remove the line below and add the "work in progress" label if you do
not want the PR to be reviewed:

@markus2330 please review my pull request

## Merging

Please add the "ready to merge" label when the build server and you say
that everything is ready to be merged.

### CODE REVIEW

**Comment by Piankero** [comment]
jenkins build libelektra please

**Comment by Piankero** [comment]
@markus2330 does the build server cache README.md documents? The build failure says it cannot find a link when I clearly removed it.

**Comment by markus2330** [comment]
Maybe there is a link to the now non-existing document?

File: doc/decisions/error_codes.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
+# Error message & handling concept
```
Comment (markus2330) [inline doc/decisions/error_codes.md]: Name of the file does not match with this heading.

File: doc/decisions/error_codes.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  The new concept should standardize errors, making it easy to categorize errors from new plugins and avoid merge conflicts.
+- No senseful way for application developers to use error codes from elektra
+  If developers of plugins/ external tools using elektra want to react to errors, they have to be very specific. At the moment there is
+  no possibility to catch all errors easily which force a certain behavior. Eg. if there happens a temporary recoverable error, developers have to
+  catch for every specific error code rather than a general hierarchical error. The new concept should make it easy to react to errors as they are
+  sensefully grouped together and are hierarchically structured.
+
+## Constraints
+
+- Error numbers must stay because they are more reliable to check against than strings
```
Comment (markus2330) [inline doc/decisions/error_codes.md]: So we keep error numbers? Why is #2686 then changing int -> char*?

File: doc/decisions/error_codes.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+- Supporting multiple programming languages
+- Plugin System
```
Comment (markus2330) [inline doc/decisions/error_codes.md]:
```suggestion
- Supporting Elektra's Plugin System
```

File: doc/decisions/error_codes.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  return codes as integers.
+- [SNMP Standard](http://www.snmp.com/protocol/):
+  Being a standard network protocol, error codes are very specific to the domain itself. A list can be found [here](https://docs.microsoft.com/en-us/windows/desktop/snmp/snmp-error-codes) and would not meet the needs of elektra at all.
+- POSIX:
+  Returning a non-zero value and retrieving the concrete information from `errno` would not suffice for elektra as it is too simple. It would not solve any of our current
+  problems like having excessive uncategorized codes for errors.
+
+## Decision
+
+All "fatal" errors will be converted to "errors" as the distinguishment is not relevant.
```
Comment (markus2330) [inline doc/decisions/error_codes.md]:
```suggestion
All "fatal" errors will be converted to "errors" as the distinction is not relevant.
```

File: doc/decisions/error_codes.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+Unused marked errors will be removed from the specification.
```
Comment (markus2330) [inline doc/decisions/error_codes.md]:
```suggestion
Unused errors will be removed from the specification.
```

File: doc/decisions/error_codes.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+Errors will be categorizes into logical groups with subgroups.
```
Comment (markus2330) [inline doc/decisions/error_codes.md]:
```suggestion
Errors will be categorized into logical groups with subgroups.
```

File: doc/decisions/error_codes.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+Each error will be made up of 5 characters, where the first 2 character indicate the highest level
+and character 3 to 5 will be used for subgrouping. Errors are prepended with the letter `C` which
+is the abbreviation for "Code".
+
+- Permanent errors
```
Comment (markus2330) [inline doc/decisions/error_codes.md]: Number missing?

File: doc/decisions/error_codes.md [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+- Validation
+  - Syntactic C04100
+  - Semantic C04200
+
+## Rationale
+
+The grouping of errors will allow developers to filter for specific as well as more general errors to correctly
+react to them programmatically.
+The new concept will permit additional subgrouping of errors in case it might be needed in the future. Imagine the case where
+"Installation" errors is too general because developers saw a need for splitting the errors in "elektra installation"
```
Comment (markus2330) [inline doc/decisions/error_codes.md]: This is now about the semantics of "Installation errors" but this document should only be about error codes?

File: doc/decisions/error_codes.md [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+
+- Permanent errors
+  - Resource
+    - Memory Allocation C01110
+    - General Resource C01120
+  - Installation C01300
+  - Logical
+    - Assertion C01410
+    - Interface C01420
+    - Broken plugin C01430
```
Comment (markus2330) [inline doc/decisions/error_codes.md]: Misleading name. What of the plugin is broken?

**Comment by Piankero** [comment]
```
error: Failed to find link to error_message.md in the file /home/jenkins/workspace/libelektra_PR-2684-I4G4C7O3FAKCIOP7INAZYROCYG3USUUVQMDO2WSLRQGR7T3U2XJA/doc/decisions/README.md

testscr_check_doc.sh RESULTS: 44 test(s) done 1 error(s).
```

There is clearly no error_message.md in the decision/README.md file anymore

File: doc/decisions/error_codes.md [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+If there is ever the need for another reaction based category, it can be extended very easily.
+
+## Implications
+
+The specification file will stay but should be untouched in most of the cases in the future. Also the C++ code generation
+file which uses the specification will stay as it is easier to change categories.
+
+Current errors will be migrated.
+
+## Related Decisions
```
Comment (markus2330) [inline doc/decisions/error_codes.md]: Links to the other decisions is missing. How do they interrelate?

File: doc/decisions/error_message_format.md [hunk 11]
```diff
@@ -0,0 +1,72 @@
+# Error message & handling concept
```
Comment (markus2330) [inline doc/decisions/error_message_format.md]: Again the title here does not match the file name (and also not the content).

File: doc/decisions/error_message_format.md [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+- Supporting multiple programming languages
+- Plugin System
+- Error Code should be preserved
+
+## Assumptions
+
+## Considered Alternatives
+
+Possible variations on what message should be displayed,
+eg. to keep the mountpoint information or on how wordings should be (with or without
```
Comment (markus2330) [inline doc/decisions/error_message_format.md]:
```suggestion
e.g., to keep the mountpoint information or on how wordings should be (with or without
```

File: doc/decisions/error_message_format.md [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+"Sorry, ...", coloring of certain parts of a message, etc.)
```
Comment (markus2330) [inline doc/decisions/error_message_format.md]: Make a list of considered alternatives.

File: doc/decisions/error_message_format.md [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+`Mountpoint`, `Configfile` in addition to the current error message.
+Furthermore a developer can use the command line argument `-d` (debug)
+to show `At` for debugging purposes.
+
+## Rationale
+
+The new error message is much more succinct which gives end users more relevant information.
+Furthermore the solution approach still holds all necessary information if requested by users.
+
+`Ingroup`, `Description` and `Module` will be removed from the error message as they provide no useful
```
Comment (markus2330) [inline doc/decisions/error_message_format.md]: This is Implication not Rationale? Description needs to be ported into Reason.

**Comment by sanssecours** [comment]
> There is clearly no error_message.md in the decision/README.md file anymore. And the other file exists

The problem here is that the file `error_message.md` still exists, while README.md contains no link to this file. You can fix this error by removing `error_message.md`.

File: doc/decisions/error_message_format.md (hunk 13 above)
Comment (Piankero) [reply doc/decisions/error_message_format.md]: The alternatives are pretty much endless on how you can rearrange an error message. I added some examples but I think it's is not feasible to list all possible variations of is possible for every field.

Comment (markus2330) [reply doc/decisions/error_message_format.md]: The list of alternatives is always endless. It is about listing the *useful* but nevertheless discarded alternatives.

**Comment by markus2330** [comment]
I have already added my review. Please carefully bring the whole docu up-to-date.

**Comment by Piankero** [comment]
> I have already added my review. Please carefully bring the whole docu up-to-date.

Beside the alternatives I changed everything you wrote. I know also wrote more alternatives

**Comment by markus2330** [comment]
Thank you.

**Pull request merged**